Merging MultiCam

BryonLewis opened the pull request:
Working on getting multi-cam merged in with the activity groups.

Similar to the Previous branch.

CameraStore.ts - CameraStore which contains a camMap: Record<string, {trackStore, groupStore}> for each camera.  The CameraStore itself has functions for accessing tracks: `getTrack, getPossibleTrack, getAnyTrack, getAllTracks`.  So now the interface into the trackStore/groupStore is through the cameraStore.

BryonLewis commented:
Got more stuff working properly.

TODO:

- [ ] Get `MultiCamTools` Panel properly functioning
- [ ] Get Camera Linking/unLinking Working.
- [ ] Make sure multicam pipelines are running properly
- [ ] Check import/export for multicam images
- [ ] Double check that single camera groups still function properly
- [ ] See how multicam-groups work and if they work properly.  There may be nuances that I haven't considered yet that could cause problems with multicam
- [ ] Implement or disable track merging in multi-camera mode
- [ ] Write some multi-camera tests.

BryonLewis commented:
TODO:

- [ ] Update documentation
- [ ] More in-depth review of groups

subdavis commented on client/dive-common/use/useSave.ts:
```diff
@@ -89,59 +104,84 @@ export default function useSave(
     pendingSaveCount.value = 0;
   }
 
+  function addCamera(cameraName: string) {
```
I find it slightly confusing that `CameraStore` has a `removeCamera` method but this file does not.

subdavis commented on client/platform/desktop/frontend/components/ViewerLoader.vue:
```diff
@@ -48,23 +43,23 @@ export default defineComponent({
             title: 'Pipeline Finished',
             text: [`Pipeline: ${currentJob.job.title}`,
-              'finished running on the current dataset.  Click reload to load the annotations.  The current annotations will be replaced with the pipeline output.',
+              'finished running sucesffully on the current dataset.  Click reload to load the annotations.  The current annotations will be replaced with the pipeline output.',
```
sucesffully

BryonLewis replied:
yeah, I don't like english.

subdavis commented on client/platform/desktop/frontend/components/ViewerLoader.vue:
```diff
@@ -128,21 +137,21 @@ export default defineComponent({
       <Export
         v-if="datasets[id]"
-        :id="compoundId"
+        :id="modifiedId"
```
Out of scope, but I think there needs to be a way to export the entire multicamera dataset, not just 1 cam at a time.

BryonLewis replied:
I think that comes down to implementing on desktop the dataset export and creation of zip files as well.

subdavis commented on client/platform/desktop/frontend/components/ViewerLoader.vue:
```diff
@@ -128,21 +137,21 @@ export default defineComponent({
     <template #title-right>
       <RunPipelineMenu
-        :selected-dataset-ids="[id]"
+        :selected-dataset-ids="[modifiedId]"
```
Since pipelines can run on multi-camera data, why does this menu take the specific identifier of an individual camera?  Should it take both?

BryonLewis replied:
The capability to run single cam pipelines on the selected Camera.  Especially stuff that doesn't modify the trackIds like polygon generation pipelines.  There is the warning currently about doing this currently without knowing what pipeline you are running.  I can put a `prop` that is selectedCamera and it just defaults to an ignored value as well.  If you want I can do that.

subdavis commented on client/src/BaseFilterControls.ts:
```diff
@@ -22,8 +21,9 @@ export interface AnnotationWithContext<T extends Track | Group> {
 
 export interface FilterControlsParams<T extends Track | Group> {
-  store: BaseAnnotationStore<T>;
+  sorted: Ref<T[]>;
   markChangesPending: () => void;
+  remove: (id: AnnotationId) => void;
```
This moves back toward the `useFilterControls` style where we passed individual bits of state and handlers.

I understand why it's necessary, I just want to avoid going down the path where all our constructor functions end up looking like the composition function params used to with half a dozen arguments each.

Nothing to do here, just something to think about.

subdavis commented on client/src/TrackStore.ts:
```diff
@@ -3,14 +3,15 @@ import BaseAnnotationStore from './BaseAnnotationStore';
-  add(frame: number, defaultType: string, afterId?: AnnotationId) {
-    const track = new Track(this.getNewId(), {
+  add(frame: number, defaultType: string, afterId?: AnnotationId, overrideTrackId?: number) {
+    const newId = overrideTrackId ?? this.getNewId();
```
right side `?? this.getNewId();` is dead code.  `trackStore.add` is only used in one place, and it provides the override.  Might as well just make it a required argument and drop support for `this.getNewId`?

BryonLewis replied:
Thanks I think this led to me catching some other bugs.  The fact that everywhere we did a trackAdd we need to make sure if overrideTrackId is underfined it needs to be set to `cameraStore.getNewTrackId()` instead of the trackstore version.  Could have conflicting trackIds then.

subdavis commented on client/src/components/GroupList.vue:
```diff
@@ -32,22 +34,34 @@ export default defineComponent({
+    const groupStoreRef = computed(() => cameraStore.camMap.get(selectedCamera.value)?.groupStore);
+    const getAnnotation = (id: AnnotationId) => {
+      if (!groupStoreRef.value) {
+        throw Error(`Could not find groupStore for Camea: ${selectedCamera.value}`);
```
Camea

subdavis reviewed (changes requested):
## Review

I think this is an improvement over the original version, it seems a bit more readable to me.  I feel like I can follow everything pretty well.

Main issue I found is that the left track list seems to show duplicates of tracks for every camera.   The tracks on the left are linked, I would expect there to only be one.

<img width="1410" alt="Screen Shot 2022-06-10 at 3 00 50 PM" src="https://user-images.githubusercontent.com/4214172/173135139-022a21f4-14ba-436e-9909-605040c197bf.png">

I also made a version of multicam without the CameraStore canary variable using `triggerRef` and `shallowRef` instead, which could likely replace a lot of our canary vars.  https://github.com/Kitware/dive/pull/1259

## General requested changes

* make `'singleCam'` a defined constant rather than just a string.

## TODO

- [ ] test all cases for pipelines
- [ ] test all variations of multicam and stereo import.
- [ ] finish line-by-line read through (I've done about 70%)

BryonLewis commented on client/src/track.ts:
```diff
@@ -188,7 +188,10 @@ export default class Track extends BaseAnnotation {
    * self if there are conflicts
    */
-  merge(others: Track[]) {
+  merge(others: Track[], disableNotifier = false) {
+    if (disableNotifier) {
+      this.notifierEnabled = false;
+    }
```
I'm ham fistedly abusing merge to get a track Object which has a combined begin/end so it can seek to the nearest frame of the track when selected using keyboard or the track list.  Maybe it would be best to refactor that out into something else and remove the `disableNotifier` stuff.

BryonLewis commented on client/src/components/controls/Controls.vue:
```diff
@@ -141,7 +143,7 @@ export default defineComponent({
             :color="mediaController.lockedCamera.value ? 'primary': 'default'"
             title="center camera on selected track"
-            @click="mediaController.toggleLockedCamera"
+            @click="mediaController.toggleLockedCamera()"
```
No I think I got confused on another project and there was something with a this context that was messed up.  For some reason it flashed in my head that this was an issue and it isn't.  I'll revert it.

BryonLewis commented:
I addressed some Comments and am not done yet.

- I still like having a full track by abusing `merge` to get it back.  I could just get the begin/end times but I need to make sure it is only used for that merging stuff.
- Not a fan of the modified `add(` requiring the overrideId from the camera Store makes it a little annoying in the tests, I may refactor it.

subdavis reviewed (changes requested):
# Possible issues

- [x] * Running single-cam pipelines on a multicam dataset doesn't put the dataset into read-only mode.
- [ ] * 2-cam suppressor for sea lions results in no annotations in the STAR camera.  I'm not sure if this is intended, or if I ran it in a weird way.  2 cam tracker seems to work as intended.
- [ ] * The tooltip for adding a link in another camera seems like it should say "Add link..." and not "Add track..." because it doesn't put you in creation mode.  The hover text for unlink is correct. 
- [ ] * The toggle keyframe button (and split button and interpolate button) seem to base their state and functions on only one camera.  If a track only exists in two cameras, these buttons may be grayed out when they shouldn't be, and if a keyframe exists on both cameras at a given frame, toggling the keyframe should either impact both cameras or the editing camera, but instead it just always affects one.
- [ ] * The timeline seems to only consider 1 camera when showing the keyframe visualization, and I can't remember if this is intended.   I totally understand not being able to display interpolation correctly, but it seems like it should be possible to show the correct start and end times of a track that spans cameras.  For example, a track that starts on the left camera at frame 0 and ends on the right camera at frame 4 will appear as a track of length 1 on frame 0 in the timeline.

    * ![Screenshot from 2022-06-21 10-49-05](https://user-images.githubusercontent.com/4214172/174829499-1358af1c-082f-4e47-9840-25b7501a0e04.png)

- [ ] * Utility pipelines do not appear to work.

    * ![Screenshot from 2022-06-21 11-12-39](https://user-images.githubusercontent.com/4214172/174834999-c9218b45-a676-48cb-869b-2e7689b58163.png)

- [ ] * The styling for the editing button mode controls seem broken.  the blue/green highlighting is not working.
- [ ] * Likewise, the editing mode indicator is broken, and always says "not editing" no matter the mode.

BryonLewis commented:
As of right now it's fixed where the keyframe/split state is representative across all of cameras.  This is done by using a mergeTrack for the sorted list provided.  This doesn't solve the instance of what to do when a keyframe is toggled.
Do I attempt to toggle all keyframes on that camera?  If split do I split all tracks at that point into individual tracks, or do I split it based on time for those tracks.

Some of this is solved (detection editing) with the multicamera tools.  Would it be better if I just had a multicam icon there that would open the multicam tools and select that specific track.  It doesn't solve the problem of toggling on interpolation.  If needed that could be something placed inside of the multicam tools as well.

BryonLewis commented:
I like the idea of using the unified star for keyframe.  I did have the controls disabled in multi-cam before.  I'll modify the track camera controls to either be in a  disables/hidden state or open the multicam tools directly.

The switch to using a merged Track for all of the sorted requirements caused the track type problems.  I forgot that in the older version a track type change would change the type across all cameras for that track.  I need to update that to fix it.

BryonLewis commented:
New bug: the side-bar context persists when switching between single and multicamera.  Meaning that if you go to as single camera dataset and then back to a multicam one it doesn't re-register the multicam data.  I need to update the group and multicam registration of the contexts to make sure they are always added/removed during the `load()` for each one.

subdavis reviewed (changes requested):
This looks really good, and it resolves all the issues I can think to test.

## Final comments

* I noticed that type styles are now persisting between annotations, which is unexpected.  So if you have the same type "fish" in 2 datasets, and you change it to be red in one, it's now red everywhere.  Changing type styles also no longer trigger a pending save badge: they just get persisted immediately.

## Things to figure out later.

* I wonder if there should be a way for users to indicate that they want to import a camera's annotations as unique or linked.  Right now, track IDs aren't deconflicted at import (which I agree is the correct default) but I wonder if we'll have people for whom they want to import annotations that were independently generated for 2 cameras that don't get linked if they both happen to have a track ID 0.
* The above DOES appear to be an issue for single-camera pipelines.   For example: if you run motion on left, then run motion on right, the two runs were independent and generate tracks starting from ID 0, so they get linked in weird ways by accident.  I don't think this is within scope to solve in this PR.

BryonLewis commented:
> * I wonder if there should be a way for users to indicate that they want to import a camera's annotations as unique or linked.  Right now, track IDs aren't deconflicted at import (which I agree is the correct default) but I wonder if we'll have people for whom they want to import annotations that were independently generated for 2 cameras that don't get linked if they both happen to have a track ID 0.
> * The above DOES appear to be an issue for single-camera pipelines.   For example: if you run motion on left, then run motion on right, the two runs were independent and generate tracks starting from ID 0, so they get linked in weird ways by accident.  I don't think this is within scope to solve in this PR.

Yeah that is the reason for the pop-up warning about running the single cameras.  The top part of what you said could relate to the idea of running additive pipelines (pipelines that create a grouping or labeling that prevents overwriting all of the current results).  So even on single camera datasets you can import multiple sets of annotations but give them either a differing type name prefix or group so you can compare.  Being able to compare two different tracker runs could be done that way.

subdavis reviewed (approved):
Spent about 30 more minutes just fiddling with everything I could think of.  I can't find any issues, it all works exactly as intended.

Great work, I know this has been a huge effort!
